### PR TITLE
File rename refactor

### LINKED ISSUE #2050: Click outside should save a renaming action
Based on me shadowing a Files user, and checking the behavior of a normal OS (windows or linux but I bet mac is the same).
When you click out after renaming a file, it actually saves it. On files it cancels the renaming action.

### PR DESCRIPTION
**Pull request by Tbaut**
closes #2050

Since the edition isn't exactly instant even with good internet, I find it good to have a spinner to show something is happening

---
Submission checklist: 

   

#### Layout
- [x] Change looks good in the desktop web ui
- [x] Change looks good in the mobile web ui

#### Theme
- [x] Components / elements inspected in light mode
- [x] Components / elements inspected in dark mode

### CODE REVIEW

**Comment by asnaith** [comment]
I think this feels more natural 👍. 

There is no way for a user to cancel once they have made their change now though. 

We also need to apply the same change to the Shared page, when renaming a Shared Folder it did not save when I clicked outside the input. 

It would be good to apply the same change to files within buckets on Storage UI too (or create a seperate issue), what do you think? 

I wasn't able to see the spinner when I throttled the connection in the dev tools.

**Comment by Tbaut** [comment]
> There is no way for a user to cancel once they have made their change now though.

There is by pressing the `esc` key, like on an os.

Thanks for checking it out, I'll apply these changes to the other areas where we rename stuff.

**Comment by Tbaut** [comment]
Storage needed some small refactor. This is being a big PR now. I tried to test all cases, and couldn't spot anything missing, feel free to play around and rename anything you can rename :)

I did not change the mobile way with a loader on the modal, just to keep things from exploding totally.

**Comment by asnaith** [comment]
Just checked out the latest changes. Share folder renaming in Files UI and file renaming in Storage UI buckets are both good now. 

I did notice on storage ui that we're not doing the same thing with file extensions when we're renaming so I'll create a separate issue as it would be good to have them behaving the same. 

I realized I was seeing the spinner before but in dark mode it's really difficult to see: 

Screen Shot 2022-03-30 at 2 55 00 PM

**Comment by tanmoyAtb** [comment]
The implementation is really great, here are some thoughts that came into my mind.
These are really borderline nitpicking though. 

1. on hitting rename, the editing field should get in focus (I start typing right after hitting rename).
2. If you clear the field, and click outside, this should cancel the rename operation (This is common OS behaviour as well, sth I noticed when I didn't want to hit `esc`).

**Comment by Tbaut** [comment]
@tanmoyAtb 
> the editing field should get in focus

Agreed, I looked for this, the `autofocus` is set properly, but there's generally a focus problem linked to the drag and drop IMHO, so formik fails at focusing the field. We should double check this with https://github.com/ChainSafe/ui-monorepo/issues/2051


> If you clear the field, and click outside, this should cancel the rename operation (This is common OS behaviour as well, sth I noticed when I didn't want to hit esc).

Ah good point! Definitely something I can do


@asnaith I hadn't tested dark mode yet (and didn't check the box in the first post :D ), it's fixed now :)
image

**Comment by asnaith** [comment]
> If you clear the field, and click outside, this should cancel the rename operation (This is common OS behaviour as well, sth I noticed when I didn't want to hit esc).

>>Ah good point! Definitely something I can do

This seems implemented already :) 

https://user-images.githubusercontent.com/3982665/161147068-72c4fb35-40ce-4c80-9c22-23f3031178a1.mov

> @asnaith I hadn't tested dark mode yet (and didn't check the box in the first post :D ), it's fixed now :)

That's why I switched to dark mode straight away 😂. Looks great now.

**Review by asnaith — APPROVED** [review]

File: packages/files-ui/src/Components/Modules/FileBrowsers/views/FileSystemItem/FileSystemTableItem.tsx [hunk 1]
```diff
@@ -178,10 +181,14 @@ const FileSystemTableItem = React.forwardRef(
       initialValues: { name: fileName },
       validationSchema: nameValidator,
       onSubmit: (values: { name: string }) => {
+
         const newName = extension !== "" ? `${values.name.trim()}.${extension}` : values.name.trim()
 
-        if (newName !== name) {
-          newName && handleRename && handleRename(file.cid, newName)
+        if (newName !== name && !!newName && handleRename) {
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/FileBrowsers/views/FileSystemItem/FileSystemTableItem.tsx:187]: yup that's in the code, if `newName` is "" it'll be falsy and go to the `else` and stop editing.

**Comment by Tbaut** [comment]
Thanks Andrew for checking, it's indeed already supported :)

**Review by tanmoyAtb — APPROVED** [review]
The rename functionality feels fully intuitive now. 💯

File: packages/files-ui/src/Components/Modules/FileBrowsers/views/FileSystemItem/FileSystemGridItem.tsx [hunk 2]
```diff
@@ -291,7 +297,13 @@ const FileSystemGridItem = React.forwardRef(
                 </Form>
               </FormikProvider>
             )
-            : <div className={classes.gridFolderName}>{name}</div>
+            : <div className={classes.gridFolderName}>
+              {name}{!isEditingLoading && <Loading
```
Comment (tanmoyAtb) [inline packages/files-ui/src/Components/Modules/FileBrowsers/views/FileSystemItem/FileSystemGridItem.tsx]: caught a small glitch
```suggestion
              {name}{isEditingLoading && <Loading
```

File: packages/files-ui/src/Components/Modules/FileBrowsers/views/FileSystemItem/FileSystemGridItem.tsx [hunk 3]
```diff
@@ -196,8 +199,11 @@ const FileSystemGridItem = React.forwardRef(
       onSubmit: (values: { name: string }) => {
         const newName = extension !== "" ? `${values.name.trim()}.${extension}` : values.name.trim()
 
-        if (newName !== name) {
-          newName && handleRename && handleRename(file.cid, newName)
+        if (newName !== name && newName && handleRename) {
```
Comment (Tbaut) [inline packages/files-ui/src/Components/Modules/FileBrowsers/views/FileSystemItem/FileSystemGridItem.tsx]: To align with the rest
```suggestion
        if (newName !== name && !!newName && handleRename) {
```

**Pull request merged**